File: oop/design_deck_of_cards copy.js
// JavaScript (ES6+)

class ICard {
  // Just a marker; no enforcement in plain JS
  stringify() {
    throw new Error("Must override stringify()");
  }
}

class PokerCard extends ICard {
  constructor(suit, value) {
    super();
    this.suit = suit;
    this.value = value;
  }
  stringify() {
    return `${this.value} of ${this.suit}s`;
  }
}

class FlinchCard extends ICard {
  constructor(number) {
    super();
    this.number = number;
  }
  stringify() {
    return `${this.number}`;
  }
}

class Deck {
  constructor() {
    this.cards = [];
  }
  create(cards) {
    this.cards = cards.slice();
  }
  take() {
    return this.cards.pop();
  }
  look() {
    return this.cards[this.cards.length - 1].stringify();
  }
}

// Usage
const cards = [new PokerCard("club", "three"), new PokerCard("heart", "four")];
const deck = new Deck();
deck.create(cards);
console.log(deck.look()); // -> "four of hearts"

File: oop/design_deck_of_cards copy.py
# Python 3

from abc import ABC, abstractmethod
from typing import List


class ICard(ABC):
    @abstractmethod
    def stringify(self) -> str:
        pass


class PokerCard(ICard):
    def __init__(self, suit: str, value: str):
        self.suit = suit
        self.value = value

    def stringify(self) -> str:
        return f"{self.value} of {self.suit}s"


class FlinchCard(ICard):
    def __init__(self, number: int):
        self.number = number

    def stringify(self) -> str:
        return str(self.number)


class Deck:
    def __init__(self):
        self.cards: List[ICard] = []

    def create(self, cards: List[ICard]) -> None:
        self.cards = list(cards)

    def take(self) -> ICard:
        return self.cards.pop()

    def look(self) -> str:
        return self.cards[-1].stringify()


if __name__ == "__main__":
    cards = [
        PokerCard("club", "three"),
        PokerCard("heart", "four")
    ]
    deck = Deck()
    deck.create(cards)
    print(deck.look())  # -> "four of hearts"
